astropy/units/core.py

```diff
@@ -461,6 +461,11 @@ class UnitConversionError(UnitsError, ValueError):
     interpreting units in terms of other units.
     """
 
+class UnitTypeError(UnitsError, TypeError):
+    """
+    Used specifically for errors in setting to units not allowed by a class.
```

Comment:
I actually don't think we need a new error for this (just use `UnitsError` with a descriptive msg), but don't feel very strongly. If you want to keep this, the description could be clearer -- I don't really understand what is there now.